### PR TITLE
Client/Server Identity validation for the audit server

### PR DESCRIPTION
**Pull request by laczau**
This PR implements hostname/IP validation in SSL certificates.

- On client side, the server certificate's SAN field (GEN_DNS, GEN_IPADD) or CN field is compared to the server's name/address
- On server side, if tls_checkpeer option is set in sudo_logsrvd.conf, the client's certificate's SAN/CN field is compared to the client's IP address. If the client cert contains a name, the validation process checks that the name resolves to the client's IP address or not.

### CODE REVIEW

File: logsrvd/logsrvd.c [hunk 1]
```diff
@@ -1264,7 +1308,7 @@ connection_closure_alloc(int sock)
  * Allocate a connection closure and send a server hello message.
  */
 static bool
-new_connection(int sock, struct sudo_event_base *base)
+new_connection(int sock, const union sockaddr_union *s_un, struct sudo_event_base *base)
```
Comment (millert) [inline logsrvd/logsrvd.c]: I would pass in const struct sockaddr *sa here.  The union is mostly just for storage in the caller.

File: logsrvd/logsrvd.h [hunk 2]
```diff
@@ -98,6 +98,7 @@ struct connection_closure {
     struct iolog_file iolog_files[IOFD_MAX];
     int iolog_dir_fd;
     int sock;
+    char ipaddr[INET6_ADDRSTRLEN];
```
Comment (millert) [inline logsrvd/logsrvd.h]: INET6_ADDRSTRLEN won't be available on non-IPv6 systems.

File: plugins/sudoers/iolog_plugin.h [hunk 3]
```diff
@@ -86,6 +86,8 @@ enum client_state {
 /* Remote connection closure, non-zero fields must come first. */
 struct client_closure {
     int sock;
+    struct sudoers_string *host;
+    char ipaddr[INET6_ADDRSTRLEN];
```
Comment (millert) [inline plugins/sudoers/iolog_plugin.h]: INET6_ADDRSTRLEN won't be present on non-IPv6 systems

File: include/hostcheck.h [hunk 4]
```diff
@@ -0,0 +1,23 @@
+
```
[the remark this answers is not in the document]
Comment (laczau) [reply include/hostcheck.h]: done

File: lib/iolog/hostcheck.c [hunk 5]
```diff
@@ -0,0 +1,313 @@
+#include "config.h"
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: done

File: lib/iolog/hostcheck.c [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+static int
+forward_lookup_match(const char *hostname, const char *ipaddr)
+{
+    int ret = 0;
+    struct addrinfo *res = NULL, *p;
+    void *addr;
+    struct sockaddr_in *ipv4;
+#if defined(HAVE_STRUCT_IN6_ADDR)
+    struct sockaddr_in6 *ipv6;
+#endif
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+#if defined(HAVE_STRUCT_IN6_ADDR)
+        } else if(p->ai_family == AF_INET6) {
+            ipv6 = (struct sockaddr_in6 *)p->ai_addr;
+            addr = &(ipv6->sin6_addr);
+#endif
+        } else {
+            goto exit;
+        }
+
+        if (inet_ntop(p->ai_family, addr, ipstr, sizeof ipstr) != 0) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+			} while (hostname_len != 0);
+		}
+		certname_s += 2;
+		certname_len -= 2;
+	}
+	/* Compare expected hostname with the DNS name */
+	if (certname_len != hostname_len) {
+		return MatchNotFound;
+	}
+    if (strcasecmp(hostname, certname_s) != 0) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+            char *san_ip = (char *) ASN1_STRING_get0_data(current_name->d.iPAddress);
+            char san_ip_str[INET6_ADDRSTRLEN];
+
+            /* IPV4 address */
+            if(current_name->d.iPAddress->length == 4) {
+                if (inet_ntop(AF_INET, san_ip, san_ip_str, INET_ADDRSTRLEN) == NULL) {
+                    result = MalformedCertificate;
+                    break;
+                }
+            /* IPV6 address */
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+            } else if (current_name->d.ip->length == 16) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: logsrvd/logsrvd.c (hunk 1 above)
Comment (laczau) [reply logsrvd/logsrvd.c]: fixed

File: logsrvd/logsrvd.h (hunk 2 above)
Comment (laczau) [reply logsrvd/logsrvd.h]: fixed

File: plugins/sudoers/iolog_plugin.h (hunk 3 above)
Comment (laczau) [reply plugins/sudoers/iolog_plugin.h]: fixed

File: lib/iolog/hostcheck.c [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+		return MalformedCertificate;
+	}
+
+	/* Compare expected hostname with the CN */
+    if (validate_name(hostname, common_name_asn1) == MatchFound) {
+		return MatchFound;
+	}
+
+    /* check if hostname in the CN field resolves to the given ip address */
+    if (resolve && forward_lookup_match(common_name_str, ipaddr)) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+	if (common_name_entry == NULL) {
+		return Error;
+	}
+
+	/* Convert the CN field to a C string */
+	common_name_asn1 = X509_NAME_ENTRY_get_data(common_name_entry);
+	if (common_name_asn1 == NULL) {
+		return Error;
+	}			
+	common_name_str = (char *) ASN1_STRING_get0_data(common_name_asn1);
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 13]
```diff
@@ -0,0 +1,313 @@
...
+        return NoSANPresent;
+    }
+    san_names_nb = sk_GENERAL_NAME_num(san_names);
+
+    /* Check each name within the extension */
+    for (i=0; i<san_names_nb; i++) {
+        const GENERAL_NAME *current_name = sk_GENERAL_NAME_value(san_names, i);
+
+        if (current_name->type == GEN_DNS) {
+            char *dns_name = (char *) ASN1_STRING_get0_data(current_name->d.dNSName);
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+
+	/* Convert the CN field to a C string */
+	common_name_asn1 = X509_NAME_ENTRY_get_data(common_name_entry);
+	if (common_name_asn1 == NULL) {
+		return Error;
+	}			
+	common_name_str = (char *) ASN1_STRING_get0_data(common_name_asn1);
+
+	/* Make sure there isn't an embedded NUL character in the CN */
+	if (ASN1_STRING_length(common_name_asn1) != (int)strlen(common_name_str)) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 15]
```diff
@@ -0,0 +1,313 @@
...
+
+    /* Check each name within the extension */
+    for (i=0; i<san_names_nb; i++) {
+        const GENERAL_NAME *current_name = sk_GENERAL_NAME_value(san_names, i);
+
+        if (current_name->type == GEN_DNS) {
+            char *dns_name = (char *) ASN1_STRING_get0_data(current_name->d.dNSName);
+
+            /* Make sure there isn't an embedded NUL character in the DNS name */
+            if ((size_t)ASN1_STRING_length(current_name->d.dNSName) != strlen(dns_name)) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 16]
```diff
@@ -0,0 +1,313 @@
...
+                result = MalformedCertificate;
+                break;
+            } else {
+                /* Compare expected hostname with the DNS name */
+                if (validate_name(hostname, current_name->d.dNSName) == MatchFound) {
+                    result = MatchFound;
+                    break;
+                }
+
+                if (resolve && forward_lookup_match(dns_name, ipaddr)) {
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 17]
```diff
@@ -0,0 +1,313 @@
...
+                    result = MatchFound;
+                    break;
+                }
+            }
+        } else if (current_name->type == GEN_IPADD) {
+            char *san_ip = (char *) ASN1_STRING_get0_data(current_name->d.iPAddress);
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 18]
```diff
@@ -0,0 +1,313 @@
...
+            char san_ip_str[INET6_ADDRSTRLEN];
```
[the remark this answers is not in the document]
Comment (laczau) [reply lib/iolog/hostcheck.c]: fixed

File: lib/iolog/hostcheck.c [hunk 19]
```diff
@@ -0,0 +1,353 @@
...
+		return MalformedCertificate;
+	}
+
+	/* Compare expected hostname with the CN */
+    if (validate_name(hostname, common_name_asn1) == MatchFound) {
+		return MatchFound;
+	}
+
+    int common_name_length = ASN1_STRING_length(common_name_asn1);
+    unsigned char *nullterm_common_name = malloc(common_name_length + 1);
```
Comment (millert) [inline lib/iolog/hostcheck.c]: Please add a check for malloc() failure.

File: lib/iolog/hostcheck.c [hunk 20]
```diff
@@ -0,0 +1,353 @@
...
+                break;
+            } else {
+                /* Compare expected hostname with the DNS name */
+                if (validate_name(hostname, current_name->d.dNSName) == MatchFound) {
+                    result = MatchFound;
+                    break;
+                }
+
+                int dns_name_length = ASN1_STRING_length(current_name->d.dNSName);
+                unsigned char *nullterm_dns_name = malloc(dns_name_length + 1);
```
Comment (millert) [inline lib/iolog/hostcheck.c]: Please add a check for malloc() failure.

Comment (laczau) [reply lib/iolog/hostcheck.c]: checked

File: lib/iolog/hostcheck.c (hunk 19 above)
Comment (laczau) [reply lib/iolog/hostcheck.c]: checked

**Pull request merged**